### ISSUE TITLE
Error relocating bpf_sk_lookup on Linux 5.19 (Ubuntu 22.10)

### ISSUE THREAD

**Issue by SikoJobs** [cilium/ebpf#850]
**Version of cilium/ebpf:**
> v0.9.3 and 5fa0dc4

**Full error message:**
> field SkLookup: program sk_lookup: apply CO-RE relocations: relocate Struct:"bpf_sk_lookup": target Func:"bpf_sk_lookup": target Func:"bpf_sk_lookup": unsized type *btf.Func

**Value of `targets`:**
> [Struct:"bpf_sk_lookup"[fields=10] Func:"bpf_sk_lookup"[static proto=FuncProto]]
https://github.com/cilium/ebpf/blob/5fa0dc4ad0ce18242dc8bcf24ceebf10431f68c8/btf/core.go#L216-L220

**Type of `target`, value of `relo.kind`:**
> `*btf.Func`, `reloFieldByteOffset`
https://github.com/cilium/ebpf/blob/5fa0dc4ad0ce18242dc8bcf24ceebf10431f68c8/btf/core.go#L247-L255

**Comment by SikoJobs** [comment]
I always switch to newer Linux distributions to acquire newer ebpf features, and I haven't seen such error on Linux 5.15 (Ubuntu 22.04) or 5.10 (Debian 11) using previous versions of cilium/ebpf. But on Ubuntu 22.10, I've tried several recent versions of cilium/ebpf and this error always exists (on bpf_sk_lookup only).

Making `targets` above keep the first member only will temporarily solve the problem.

**Comment by lmb** [comment]
Interesting! It looks like CO-RE is trying to relocate a struct against a function, both named `bpf_sk_lookup`. Relocating against a function doesn't really make sense, which is why that unsized error pops up. Seems like [both types exist](https://elixir.bootlin.com/linux/v5.15.77/A/ident/bpf_sk_lookup) in the kernel, going back to 5.15 or earlier.

Are you using`struct bpf_sk_lookup` from bpf.h in your code? Or are you using `vmlinux.h`? It'd be great if you could provide me with a reproducer somewhere.

**Comment by SikoJobs** [comment]
> Interesting! It looks like CO-RE is trying to relocate a struct against a function, both named `bpf_sk_lookup`. Relocating against a function doesn't really make sense, which is why that unsized error pops up. Seems like [both types exist](https://elixir.bootlin.com/linux/v5.15.77/A/ident/bpf_sk_lookup) in the kernel, going back to 5.15 or earlier.

Guessed so... But it's weird that this error doesn't occur on Linux 5.15 or earlier. Here's my current solution:
```go
target := Copy(targets[i], UnderlyingType)
if _, ok := target.(*Func); ok {
	continue
}
```

> Are you using`struct bpf_sk_lookup` from bpf.h in your code? Or are you using `vmlinux.h`? It'd be great if you could provide me with a reproducer somewhere.

I'm using `vmlinux.h` generated by `bpftool btf dump file /sys/kernel/btf/vmlinux format c > vmlinux.h` on Ubuntu 22.10:
> Linux hostname 5.19.0-23-generic #24-Ubuntu SMP PREEMPT_DYNAMIC Fri Oct 14 15:39:57 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
go version go1.19.3 linux/amd64

BTW, there's no such error when using libbpf in user space.

**Comment by lmb** [comment]
> I'm using `vmlinux.h` generated by `bpftool btf dump file /sys/kernel/btf/vmlinux format c > vmlinux.h` on Ubuntu 22.10:

Ok, the reason your code is getting CO-RE relocated in the first place is because of `vmlinux.h`. As a work around, you can define `BPF_NO_PRESERVE_ACCESS_INDEX` before including `vmlinux.h` to avoid CO-RE in that case.

**Comment by SikoJobs** [comment]
> Ok, the reason your code is getting CO-RE relocated in the first place is because of `vmlinux.h`. As a work around, you can define `BPF_NO_PRESERVE_ACCESS_INDEX` before including `vmlinux.h` to avoid CO-RE in that case.

Thanks for your advice, but my program is quiet complex that requires CO-RE.

I found that `/sys/kernel/btf/vmlinux` contains `STRUCT 'bpf_sk_lookup'` and `FUNC 'bpf_sk_lookup'` on both Ubuntu 22.10 and Debian 11, so it's still unclear why this error only occurs on Ubuntu 22.10, at least in my environment.

**Comment by SikoJobs** [comment]
Okay, I did more tests and figured out what's going on here:
- On Debian 11, `/sys/kernel/btf/vmlinux` contains `FUNC 'bpf_sk_lookup'`, but my program back then didn't trigger CO-RE relocation. When I do some changes to my current program and run it on Debian 11, this error occurs.
- On Ubuntu 22.04, `/sys/kernel/btf/vmlinux` **doesn't** contain `FUNC 'bpf_sk_lookup'`, so this error didn't show up, even though my program back then had grown more complex and triggered CO-RE relocation.
- On Ubuntu 22.10, `/sys/kernel/btf/vmlinux` contains `FUNC 'bpf_sk_lookup'`, so my current program triggered this error.

In conclusion, it should be a bug of cilium/ebpf (I've tested v0.9.3, v0.9.2, v0.9.1, v0.9.0, v0.8.1, v0.8.0). For example, this error always occurs on `bpf_sk_lookup` when `/sys/kernel/btf/vmlinux` contains `FUNC 'bpf_sk_lookup'` and CO-RE relocation is triggered, which should be able to be reproduced easily. Removing type `*btf.Func` from `targets` above will fix this issue.

**Comment by SikoJobs** [comment]
Will be fixed by https://github.com/cilium/ebpf/pull/852

**Comment by lmb** [comment]
Thanks for the detailed bug report!
